espnet2/enh/loss/wrappers/mixit_solver.py

```diff
@@ -0,0 +1,85 @@
+import itertools
+
+import torch
+
+from espnet2.enh.loss.criterions.abs_loss import AbsEnhLoss
+from espnet2.enh.loss.wrappers.abs_wrapper import AbsLossWrapper
+
+
+class MixITSolver(AbsLossWrapper):
+    def __init__(
+        self,
+        criterion: AbsEnhLoss,
+        weight: float = 1.0,
+    ):
+        """Mixture Invariant Training Solver.
+
+        Args:
+            criterion (AbsEnhLoss): an instance of AbsEnhLoss
+            weight (float): weight (between 0 and 1) of current loss
+                for multi-task learning.
+        """
+        super().__init__()
+        self.criterion = criterion
+        self.weight = weight
+
+    @property
+    def name(self):
+        return "mixit"
+
+    def forward(self, ref, inf, others={}):
+        """MixIT solver.
+
+        Args:
+            ref (List[torch.Tensor]): [(batch, ...), ...] x n_spk
+            inf (List[torch.Tensor]): [(batch, ...), ...] x n_est
+        Returns:
+            loss: (torch.Tensor): minimum loss with the best permutation
+            stats: dict, for collecting training status
+            others: dict, in this PIT solver, permutation order will be returned
+        """
+        num_inf = len(inf)
+        num_ref = num_inf // 2
+        device = ref[0].device
+
+        ref_tensor = torch.stack(ref[:num_ref], dim=1)  # (batch, num_ref, ...)
+        inf_tensor = torch.stack(inf, dim=1)  # (batch, num_inf, ...)
```

Comment:
Does this solver only accept waveforms as input? I think a solver should be independent of the domain that a signal belongs to. Instead, the criterion defines whether time-domain or frequency-domain is expected.

Reply:
Considering that complex-valued spectra may be used as input, we may be careful about the data type (e.g., ComplexTensor vs torch.Tensor).

Reply:
I added the support for more input tensor shapes. The main step is in the `einsum` computation. But basically, timedomain is preferred due to the additivity in MixIT. Is there any special care necessary for ComplexTensor?

Reply:
Yes, for ComplexTensor, `torch_complex.functional.stack` should be used instead of `torch.stack`, and `FC.einsum` should be used instead of `torch.einsum`.
> You could check the data type by `isinstance(c, ComplexTensor)`.

A simple way to automatically handle the data type issue can be like:
```python
from espnet2.enh.layers.complex_utils import einsum, stack

...
stack(...)
einsum(...)
```

Reply:
It might be better to add more unit tests to cover the case where the input is a complex-valued spectrum.

Reply:
Now complex is supported.